h2o-core/src/main/java/hex/grid/filter/KeepOnlyFirstMatchFilterFunction.java

```diff
@@ -0,0 +1,47 @@
+package hex.grid.filter;
+
+import java.util.Map;
+import java.util.function.Function;
+
+/**
+ * This is a wrapper class for filtering function to make it possible to keep state.
+ * This class is intended to be used to reduce number of grid items grid search process is going to evaluate.
+ * Use this class to mark group of grid items that are equivalent in terms of their evaluation values and to only evaluate a single representative.
+ * 
+ *  Usage: pass the function to the constructor which will return `true` for the grid items of type <Map<String, Object>
+ *  that we will want to consider equal.
+ */
+public class KeepOnlyFirstMatchFilterFunction implements Function<Map<String, Object>, Boolean>{
+
+  // Function that should return `true` for grid items that we want to evaluate only one representative from.
+  public final Function<Map<String, Object>, Boolean> _baseMatchFunction;
+  private int _maxNumberOfApplications = 1;
```

Comment:
I have the feeling this could be simpler.
```
private boolean _applied = false;
  @Override
  public Boolean apply(Map<String, Object> stringObjectMap) {
    Boolean applyResult = !applied && _baseMatchFunction.apply(stringObjectMap);
    _applied = true;
    return applyResult ;
  }
```
no counter needed, and the application is simply skipped if already applied.

Reply:
@seb-h2o 
> no counter needed

`_applied` is the same counter in essence. I wanted to use numbers as it reflects naming of the class `KeepOnlyFirstMatch`. And changing name of the class to something `...Applied...`  would be too general ( as any function is about application). So this was done on purpose.

> no class cast in the calling code

In order for this to be true, we would need to make all our filter functions stateful -> create wrapper classes for them so that they can have `_applied` field. I did not want to do so as I believe in most cases we don't need state for functions. So I considered adding class casting is not that expensive as creating extra classes/wrappers for all type of functions. 

>  the application is simply skipped if already applied

this is the main point of my original design. The fact that I should mark function as "already  applied` only from outside of the function itself made me change logic ( see below for details).

I think it needs to be clarified here  that we need to distinguish two things:
1) filter is applicable ( matches the condition)
2) filter was considered as applied. Example: as we apply functions sequentially, the first one, let's say `KeepOnlyFirstMatch`, matched and we decided to evaluate permutation and mark function as being applied once. But the next filter function might mark this permutation as `to be sipped` and it would brake the original intention  - to evaluate exactly once what matched `KeepOnlyFirstMatch` filter function. We would have to go `back` and "demark"  `KeepOnlyFirstMatch` filter function.

Let's take a look now at your example. It of course looks simpler but it would not work for couple of reasons( some of them already explained above).
1) `_applied` would not be set to true. Let's say permutation has `blending=false` -> `_baseMatchFunction.apply(stringObjectMap)` will return `true`. By design of `KeepOnlyFirstMatch` filter we return `true` for those permutations that we want to consider equivalent and evaluate only one representative permutation. 
```
_applied |= applyResult
```
would set `_applied`  to `false` as if we did not have a match. 
Then you return `applyResult` which is `true`. On the caller side we will negate it as usual logic for filtering out is to skip elements in the collection that return `false`. As a result, permutation will be evaluated(which is fine for first time) and `_applied` would not be set to `true`,  which will result in that we will keep evaluating permutations where `_blending = false`.
 PS: but this could be fixed somehow, for example:
```
private boolean _applied = false;

  @Override
  public Boolean apply(Map<String, Object> stringObjectMap) { 
    Boolean applyResult = !applied && ! _baseMatchFunction.apply(stringObjectMap);
     if(_applied) return false; 
      else { _applied |= applyResult;  return _baseMatchFunction.apply(stringObjectMap) ;
  }
```
Even though there is a duplication of `_baseMatchFunction.apply` it is more efficient, as we call function twice only on first iteration but them we just skip. This is actually what could be improved in my code as I incidentally got rid of duplicated code by applying automatic boolean logic refactoring feature in IDEA.

2) Let's assume we managed to set the flag `_applied` to be `true` after we run into permutation that matches `KeepOnlyFirstMatch`'s condition. In your example we return `applyResult` (which is `true`) and it means we will not mark it as skipped. But next function in the sequence of filter functions might mark it as `to be skipped`. It means we will not have a single evaluated permutation with `blending=false`. That is why I needed this post processing.

So I don't see how your suggestion could be fully applied.

Reply:
@deil87 please make an effort to write shorter replies. 
This small essay was a bit discouraging: at first sight, I immediately postponed my reading, then I forgot...
My proposal was wrong (I should have double-checked), just show a use case where it fails, thx/
Although I still think we could do better than using this manual call to `decrementCounter`.

Reply:
@seb-h2o my first answers to the questions related to this topic were shorter. I just realised that subject might be difficult for the reviewers and decided explain it with your example/suggestion. Also sometimes providing example might take more time then explanation with words. Anyway usually I would expect appreciation for detailed answers even if they are a bit long :) I will try to take into account your preferences for the future ;) 

> Although I still think we could do better than using this manual call to decrementCounter

I'm open for suggestions. IMHO as we can decide on whether function was really applied only from a higher level perspective  (based on results of application of all functions) - we will need to somehow pass this knowledge manually.